tb_plugin/test/test_profiler.py

```diff
@@ -0,0 +1,437 @@
+    # In this case, all the operators except the top operator should be removed,
+    # and all runtime/kernels belong to the children operators should be attached to the only kept one.
+    # This behavior is to keep consistent with _remove_dup_nodes in torch/autograd/profiler.py.
+    def test_remove_dup_nodes(self):
+        json_content = """
+          [{
+            "ph": "X", "cat": "Operator", 
+            "name": "aten::mm", "pid": 13721, "tid": "456",
+            "ts": 100, "dur": 100,
+            "args": {"Input dims": [], "External id": 2}
+          },
+          {
+            "ph": "X", "cat": "Operator", 
+            "name": "aten::mm", "pid": 13721, "tid": "456",
+            "ts": 110, "dur": 80,
+            "args": {"Input dims": [], "External id": 3}
+          },
+          {
+            "ph": "X", "cat": "Operator", 
+            "name": "aten::mm", "pid": 13721, "tid": "456",
+            "ts": 120, "dur": 60,
+            "args": {"Input dims": [], "External id": 4}
+          },
+          {
+            "ph": "X", "cat": "Runtime", 
+            "name": "cudaLaunchKernel", "pid": 13721, "tid": "456",
+            "ts": 130, "dur": 20,
+            "args": {"correlation": 335, "external id": 4}
+          },
+          {
+            "ph": "X", "cat": "Kernel", 
+            "name": "void gemmSN_TN_kernel_64addr", "pid": 0, "tid": "stream 7",
+            "ts": 220, "dur": 8,
+            "args": {"correlation": 335, "external id": 4}
+          }]
+        """
+        profile = parse_json_trace(json_content)
+        profile.process()
+        self.assertEqual(len(profile.op_list_groupby_name), 1)
+        self.assertEqual(profile.op_list_groupby_name[0].self_device_duration, 8)
+
+    # Test Runtime with "external id" 0.
+    # This kind of Runtime should not be attached to any operator,
+    # and should be included in accumulating device time.
+    def test_top_level_runtime(self):
+        # This operator is different thread with the runtime.
+        json_content = """
+          [{
+            "ph": "X", "cat": "Operator", 
+            "name": "aten::mm", "pid": 13721, "tid": "123",
+            "ts": 100, "dur": 100,
+            "args": {"Input dims": [], "External id": 2}
+          },
+          {
+            "ph": "X", "cat": "Runtime", 
+            "name": "cudaLaunchKernel", "pid": 13721, "tid": "456",
+            "ts": 130, "dur": 20,
+            "args": {"correlation": 335, "external id": 0}
+          },
+          {
+            "ph": "X", "cat": "Kernel", 
+            "name": "void gemmSN_TN_kernel_64addr", "pid": 0, "tid": "stream 7",
+            "ts": 220, "dur": 8,
+            "args": {"correlation": 335, "external id": 0}
+          }]
+        """
+        profile = parse_json_trace(json_content)
+        profile.process()
+        self.assertEqual(profile.op_list_groupby_name[0].device_duration, 0)
+        self.assertEqual(profile.op_list_groupby_name[0].self_device_duration, 0)
```

Comment:
Should this also test the accumulating device time?

Reply:
In this test case, the Runtime is not invoked by the Operator. So the device durations of this Operator are 0.

Reply:
> Should this also test the accumulating device time?

In this test case, the Runtime is not invoked by the Operator(because external id not equal). So the device durations of this Operator are 0.